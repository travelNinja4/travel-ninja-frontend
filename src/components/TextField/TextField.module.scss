/**
------------------------------------------------------
styles specific to the TextField component
------------------------------------------------------
**/

.container {
  display: flex;
  flex-direction: column;
}

.label {
  font-weight: 600;
  font-size: 0.95rem;
  color: var(--color-gray-800);
  margin-bottom: 0.5rem;
}

.inputField {
  font-size: 1rem;
  padding: 0.75rem 0.87rem;
  border-radius: 8px;
  border: 1px solid var(--color-gray-300);
  background-color: var(--color-white);
  color: var(--color-black);
  color-scheme: light;

  &:focus {
    outline: none;
    border-color: var(--color-indigo);
    box-shadow: 0 0 0 0.2rem var(--color-indigo-alpha-25);
  }

  &:disabled {
    background-color: var(--color-gray-50);
    color: var(--color-gray-400);
    cursor: not-allowed;
  }
}

.errorText {
  color: var(--color-red-500);
  font-size: 0.875rem;
  margin-top: 4px;
}
