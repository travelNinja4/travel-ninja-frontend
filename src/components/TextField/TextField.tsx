/**
 * TextField is a reusable input component that standardizes text input styles, behavior, and validation across the application.
 *
 * @example
 * ```tsx
 * import TextField from '@src/components/TextField'
 *
 * export default function TextField() {
 *   return (
 *          <TextField
 *            label="Username"
 *            name="username"
 *            placeholder="Enter your username"
 *            value={username}
 *            onChange={(e) => setUsername(e.target.value)}
 *            error={formErrors.username}
 *      />
 *    )
 * }
 * ```
 */

import Typography from '../Typography';
import styles from './TextField.module.scss';

/**
 * Define the props available for the TextField component.
 */
interface TextFieldProps extends React.InputHTMLAttributes<HTMLInputElement> {
  /**
   * The unique identifier for the input element.
   * Useful for associating the input with a label via the `htmlFor` attribute.
   */
  id?: string;

  /**
   * The name of the input field.
   * Used for form submissions and identifying form data.
   */
  name?: string;

  /**
   * The text label displayed alongside the input.
   * Helps describe the purpose of the input for users and accessibility tools.
   */
  label?: string | React.ReactNode;

  /**
   * The type of input (e.g., "text", "password", "email").
   * Defaults to `"text"` if not provided.
   */
  type?: string;

  /**
   * The error message to display when validation fails.
   * Typically shown below the input in a distinct color (e.g., red).
   */
  error?: string;

  /**
   * The placeholder text displayed when the input is empty.
   * Gives users a hint about what to enter.
   */
  placeholder?: string;

  /**
   * The current value of the input.
   * Used in controlled components to manage state.
   */
  value?: string;

  /**
   * Whether the input should be disabled.
   * A disabled input cannot be focused or modified by the user.
   */
  disabled?: boolean;

  /**
   * Additional CSS class names for custom styling.
   */
  className?: string;

  /**
   * The default value of the input (uncontrolled component).
   * Use `value` instead for controlled components.
   */
  defaultValue?: string;

  /**
   * Whether the input is required before form submission.
   * Adds `required` validation in HTML.
   */
  required?: boolean;

  /**
   * The maximum number of characters the user can enter.
   */
  maxLength?: number;

  /**
   * The minimum number of characters required for the input to be valid.
   */
  minLength?: number;

  /**
   * Controls capitalization behavior (e.g., "none", "sentences", "words", "characters").
   * Primarily used in mobile browsers.
   */
  autoCapitalize?: string;

  /**
   * Event handler triggered when the input value changes.
   * @param event - The change event containing the updated value.
   */
  onChange?: React.ChangeEventHandler<HTMLInputElement>;
}

export default function TextField({
  label,
  error,
  type = 'text',
  id,
  name,
  placeholder,
  value,
  disabled = false,
  className,
  defaultValue,
  required = false,
  maxLength,
  minLength,
  autoCapitalize,
  onChange,
  ...rest
}: TextFieldProps) {
  return (
    <div data-testid="TextFieldTest" className={styles.container}>
      {label && (
        <Typography tag="label" htmlFor={id} className={styles.label}>
          {label}{' '}
          {required && (
            <Typography tag="span" className={styles.label}>
              *
            </Typography>
          )}
        </Typography>
      )}
      <input
        type={type}
        id={id}
        name={name}
        placeholder={placeholder}
        value={value}
        className={styles.inputField}
        disabled={disabled}
        defaultValue={defaultValue}
        required={required}
        maxLength={maxLength}
        minLength={minLength}
        autoCapitalize={autoCapitalize}
        onChange={onChange}
        {...rest}
      />
      {error && <p className={styles.errorText}>{error}</p>}
    </div>
  );
}
