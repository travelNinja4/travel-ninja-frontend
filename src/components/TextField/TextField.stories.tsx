import type { Meta, StoryObj } from '@storybook/nextjs-vite';
/**
import { fn } from 'storybook/test';
*/
import TextField from './TextField';

const meta = {
  title: 'components/TextField',
  component: TextField,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    label: {
      control: 'text',
      description: 'Label displayed above the input',
      defaultValue: 'Sample Label',
    },
    type: {
      control: 'text',
      description: 'HTML input type',
      defaultValue: 'text',
    },
    placeholder: {
      control: 'text',
      description: 'Placeholder text for the input',
      defaultValue: 'Enter value',
    },
    error: {
      control: 'text',
      description: 'Error message to display below the input',
    },
    disabled: {
      control: 'boolean',
      description: 'Whether the input is disabled',
      defaultValue: false,
    },
    required: {
      control: 'boolean',
      description: 'Whether the input is required',
      defaultValue: false,
    },
    value: {
      control: 'text',
      description: 'Controlled value of the input',
    },
    onChange: {
      action: 'changed',
      description: 'Callback when input value changes',
    },
  },
  args: {
    /** onClick: fn() **/
  },
} satisfies Meta<typeof TextField>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    label: 'Username',
    placeholder: 'Enter your username',
    required: true,
  },
};

export const WithError: Story = {
  args: {
    label: 'Email',
    placeholder: 'Enter your email',
    error: 'Invalid email address',
  },
};

export const Disabled: Story = {
  args: {
    label: 'Disabled Field',
    placeholder: 'You cannot type here',
    disabled: true,
  },
};
