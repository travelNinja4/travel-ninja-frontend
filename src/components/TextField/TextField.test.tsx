import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import TextField from './TextField';

describe('Home', () => {
  /** Base props for TextField **/
  const baseProps = {
    label: 'TextField Test',
    name: 'testField',
    id: 'testFieldId',
    placeholder: 'Enter text here',
    value: '',
    onChange: jest.fn(),
  };

  /**
   * @test
   * Renders the TextField component without throwing an error
   * and checks if the TextFieldTest text is present in the document.
   *
   * This ensures that the component loads correctly with the base props.
   */
  it('renders without crashing', () => {
    render(<TextField {...baseProps} />);
    expect(screen.getByTestId('TextFieldTest')).toBeInTheDocument();
  });

  /**
   * @test
   * Ensures that the TextField component renders without throwing any runtime errors.
   *
   * This is a safety check to confirm that the component can mount successfully
   * with the provided base props, even if it does not render any visible content.
   */
  it('renders without throwing', () => {
    expect(() => render(<TextField {...baseProps} />)).not.toThrow();
  });

  /**
   * @test
   * Captures a snapshot of the TextField component with base props
   * and ensures it matches the previously stored snapshot.
   *
   * This helps detect any unexpected UI changes.
   */
  it('matches snapshot', () => {
    const { asFragment } = render(<TextField {...baseProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
