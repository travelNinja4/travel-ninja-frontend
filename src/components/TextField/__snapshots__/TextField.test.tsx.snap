// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Home matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
    data-testid="TextFieldTest"
  >
    <label
      class="typography label align-left label"
      data-testid="TypographyTest"
      for="testFieldId"
    >
      TextField Test 
    </label>
    <input
      class="inputField"
      id="testFieldId"
      name="testField"
      placeholder="Enter text here"
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;
